File: src/main/java/at/jku/ssw/coco/intellij/psi/impl/CocoPsiImplUtil.java
package at.jku.ssw.coco.intellij.psi.impl;

import at.jku.ssw.coco.intellij.CocoIcons;
import at.jku.ssw.coco.intellij.psi.*;
import com.intellij.lang.ASTNode;
import com.intellij.navigation.ItemPresentation;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiNamedElement;
import org.apache.commons.lang.NotImplementedException;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

/**
 * @author Thomas Scheinecker <a href="mailto:tscheinecker@gmail.com">tscheinecker@gmail.com</a>
 */
public class CocoPsiImplUtil {


    public static String getName(PsiNamedElement element) {
        ASTNode nameNode = element.getNode().findChildByType(CocoTypes.IDENT);
        return nameNode != null ? nameNode.getText() : null;
    }

    public static PsiElement setName(PsiNamedElement element, String newName) {
        ASTNode nameNode = element.getNode().findChildByType(CocoTypes.IDENT);
        if (nameNode != null) {

            if (element instanceof CocoCompiler) {
                CocoCompiler compiler = CocoElementFactory.createCompiler(element.getProject(), newName);
                ASTNode newNameNode = compiler.getNode().findChildByType(CocoTypes.IDENT);
                assert newNameNode != null;
                element.getNode().replaceChild(nameNode, newNameNode);
            } else {
                throw new NotImplementedException();
            }

        }
        return element;
    }

    public static PsiElement getNameIdentifier(PsiNamedElement element) {
        ASTNode nameNode = element.getNode().findChildByType(CocoTypes.IDENT);
        if (nameNode != null) {
            return nameNode.getPsi();
        } else {
            return null;
        }
    }

    public static ItemPresentation getPresentation(final PsiNamedElement element) {
        return new ItemPresentation() {
            @Nullable
            @Override
            public String getPresentableText() {
                return element.getName();
            }

            @Nullable
            @Override
            public String getLocationString() {
                return element.getContainingFile().getName();
            }

            @Nullable
            @Override
            public Icon getIcon(boolean unused) {
                return CocoIcons.FILE;
            }
        };
    }

    @Nullable
    public static String getCharacterReferenceName(@NotNull HasCocoCharacterReference element) {
        return getIdentText(element);
    }

    @Nullable
    public static String getTokenReferenceName(@NotNull HasCocoTokenReference element) {
        return getIdentText(element);
    }

    @Nullable
    public static String getProductionReferenceName(@NotNull HasCocoProductionReference element) {
        return getIdentText(element);
    }

    @Nullable
    public static String getIdentText(@NotNull HasIdent element) {
        PsiElement ident = element.getIdent();
        if (ident != null) {
            return ident.getText();
        }

        return null;
    }
}

File: src/main/kotlin/me/salzinger/intellij/coco/psi/impl/CocoPsiImplUtil.kt
package me.salzinger.intellij.coco.psi.impl

import com.intellij.navigation.ItemPresentation
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNameIdentifierOwner
import com.intellij.psi.PsiNamedElement
import com.intellij.psi.impl.source.tree.LeafPsiElement
import me.salzinger.intellij.coco.CocoIcons
import javax.swing.Icon

/**
 * @author Thomas Salzinger [tsalzinger@gmail.com](mailto:tsalzinger@gmail.com)
 */
object CocoPsiImplUtil {


    @JvmStatic
    fun getName(element: PsiNameIdentifierOwner): String? {
        val nameIdentifier = element.nameIdentifier
        return nameIdentifier?.text
    }

    @JvmStatic
    fun setName(element: PsiNameIdentifierOwner, newName: String): PsiElement {
        val nameIdentifier = element.nameIdentifier

        (nameIdentifier as? LeafPsiElement)?.replaceWithText(newName) ?: throw UnsupportedOperationException("Cannot rename element of type " + element.javaClass.simpleName)

        return element
    }

    @JvmStatic
    fun getNameIdentifier(element: PsiNamedElement): PsiElement? {
        val nameNode = element.node.findChildByType(me.salzinger.intellij.coco.psi.CocoTypes.IDENT)
        return nameNode?.psi
    }

    @JvmStatic
    fun getTextOffset(cocoCompiler: me.salzinger.intellij.coco.psi.CocoCompiler): Int {
        return cocoCompiler.startOffsetInParent + cocoCompiler.nameIdentifier!!.startOffsetInParent
    }

    @JvmStatic
    fun getPresentation(element: PsiNamedElement): ItemPresentation {
        return object : ItemPresentation {
            override fun getPresentableText(): String? {
                return element.name
            }

            override fun getLocationString(): String? {
                return element.containingFile.name
            }

            override fun getIcon(unused: Boolean): Icon? {
                return CocoIcons.FILE
            }
        }
    }

    @JvmStatic
    fun getPresentation(element: me.salzinger.intellij.coco.psi.CocoDirective): ItemPresentation {
        return object : ItemPresentation {
            override fun getPresentableText(): String? {
                return element.text
            }

            override fun getLocationString(): String? {
                return element.containingFile.name
            }

            override fun getIcon(unused: Boolean): Icon? {
                return CocoIcons.FILE
            }
        }
    }
}
